pkg/util/metrics/registry.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"github.com/blang/semver"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+var (
+	// todo: load the version dynamically at application boot.
```

Comment:
Is there actually a way to do this reliably? Is this not something that is injected as compiler flags? I don't know enough about this, but we should investigate this space deeper before we move forward with the assumption that this works.

Reply:
We load them through ldflags today for the apimachinery binary so it already works. I would need to make sure they get loaded for the other binaries as well.

Reply:
Ideally I would like genericize the way we do version loading across kubernetes binaries. The struct we use to represent version data in apimachinery is very apimachinery specific.

Reply:
Right I'm saying how much does Kubernetes already force people into setting ldflags appropriately and is that something we can rely on.

Reply:
It's automated as a part of the build process. We even have fallbacks in place for deriving this information from git (though, as I understand it, is only necessary for certain dev build flows).

Reply:
This seems somewhat brittle. I don’t think Kubernetes support is bound to complying to a certain build process or the assumption that people use the “official” release builds. I’m saying if we move forward with this, we may end up mandating that.

Reply:
I think your idea of mandating that is a good idea, it seems like the safest and probably most prudent thing to do. Looking over the build scripts it seems like it should currently always be there during build time, but the guarantee would be nice. 

I've removed the hardcoded version information in a recent commit and just parse out the version information which is populated during build. It's a little short around test cases currently, but I will supplement them if this seems okay-ish with you. lmkwyt.